File: src/api/go/pxapi/types/schema.go
/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package types

// TableMetadata contains the table metadata state.
type TableMetadata struct {
	// Name of the TableMetadata.
	Name string
	// ColInfo has the array index schemas of each column.
	ColInfo []ColSchema
	// ColIdxByName has the mapping from column names to column indices.
	ColIdxByName map[string]int64
}

// IndexOf returns the index of a column by name. -1 is returned if the column does not exist.
func (t *TableMetadata) IndexOf(colName string) int64 {
	idx, ok := t.ColIdxByName[colName]
	if !ok {
		return -1
	}
	return idx
}

// ColSchema has the per column schema.
type ColSchema struct {
	// Name of the column.
	Name string
	// Type of the column.
	Type DataType
	// SemanticType of the column.
	SemanticType SemanticType
}

// Record stores information about a single record.
type Record struct {
	// Data is the array index type erased values.
	Data []Datum
	// TableMetadata stores a pointer to underlying table metadata.
	TableMetadata *TableMetadata
}

// GetDatum returns the value of the given column.
func (r *Record) GetDatum(colName string) Datum {
	idx := r.TableMetadata.IndexOf(colName)
	if idx < 0 {
		return nil
	}
	return r.Data[idx]
}

// GetDatumByIdx returns the data at a given column index.
func (r *Record) GetDatumByIdx(idx int64) Datum {
	return r.Data[idx]
}

File: src/table_store/schema/schema.cc
/*
 * Copyright 2018- The Pixie Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <algorithm>
#include <vector>

#include <absl/strings/str_format.h>
#include "src/common/base/base.h"
#include "src/table_store/schema/schema.h"

namespace px {
namespace table_store {
namespace schema {

bool Schema::HasRelation(int64_t id) const { return relations_.find(id) != relations_.end(); }

std::vector<int64_t> Schema::GetIDs() const {
  std::vector<int64_t> ids(relations_.size());
  std::transform(relations_.begin(), relations_.end(), ids.begin(),
                 [](const auto& pair) { return pair.first; });
  return ids;
}

void Schema::AddRelation(int64_t id, const Relation& relation) {
  VLOG_IF(1, HasRelation(id)) << absl::StrFormat("WARNING: Relation %d already exists", id);
  relations_[id] = relation;
}

std::string Schema::DebugString() const {
  if (relations_.empty()) {
    return "Relation: <empty>";
  }
  std::string debug_string = "Relation:\n";
  for (const auto& pair : relations_) {
    debug_string += absl::StrFormat("  {%d} : %s\n", pair.first, pair.second.DebugString());
  }
  return debug_string;
}

StatusOr<const Relation> Schema::GetRelation(int64_t id) const {
  if (!HasRelation(id)) {
    return error::NotFound("no such relation: $0", id);
  }
  return relations_.at(id);
}

Status Schema::ToProto(schemapb::Schema* schema,
                       const absl::flat_hash_map<std::string, schema::Relation>& relation_map) {
  CHECK(schema != nullptr);
  auto map = schema->mutable_relation_map();
  for (auto& [table_name, relation] : relation_map) {
    schemapb::Relation* relation_pb = &(*map)[table_name];
    PX_RETURN_IF_ERROR(relation.ToProto(relation_pb));
  }
  return Status::OK();
}

}  // namespace schema
}  // namespace table_store
}  // namespace px
